File: ProjectTest/OneLang/CSharp/One/Transforms/CollectInheritanceInfo.cs
using One.Ast;

namespace One.Transforms
{
    public class CollectInheritanceInfo {
        public void visitClass(Class cls) {
            var allBaseIIntfs = cls.getAllBaseInterfaces();
            var intfs = allBaseIIntfs.map(x => x is Interface int_ ? int_ : null).filter(x => x != null);
            var clses = allBaseIIntfs.map(x => x is Class class_ ? class_ : null).filter(x => x != null && x != cls);
            
            foreach (var field in cls.fields)
                field.interfaceDeclarations = intfs.map(x => x.fields.find(f => f.name == field.name)).filter(x => x != null);
            
            foreach (var method in cls.methods) {
                method.interfaceDeclarations = intfs.map(x => x.methods.find(m => m.name == method.name)).filter(x => x != null);
                method.overrides = clses.map(x => x.methods.find(m => m.name == method.name)).find(x => x != null);
                if (method.overrides != null)
                    method.overrides.overriddenBy.push(method);
            }
        }
        
        public void visitPackage(Package pkg) {
            foreach (var file in Object.values(pkg.files))
                foreach (var cls in file.classes)
                    this.visitClass(cls);
        }
    }
}

File: ProjectTest/OneLang/Java/src/main/java/OneLang/One/Transforms/CollectInheritanceInfo/CollectInheritanceInfo.java
import java.util.Arrays;

public class CollectInheritanceInfo implements ITransformer {
    String name = "CollectInheritanceInfo";
    public String getName() { return this.name; }
    public void setName(String value) { this.name = value; }
    
    public CollectInheritanceInfo()
    {
        // C# fix
        this.setName("CollectInheritanceInfo");
    }
    
    public void visitClass(Class cls) {
        var allBaseIIntfs = cls.getAllBaseInterfaces();
        var intfs = Arrays.stream(Arrays.stream(allBaseIIntfs).map(x -> x instanceof Interface ? ((Interface)x) : null).toArray(Interface[]::new)).filter(x -> x != null).toArray(Interface[]::new);
        var clses = Arrays.stream(Arrays.stream(allBaseIIntfs).map(x -> x instanceof Class ? ((Class)x) : null).toArray(Class[]::new)).filter(x -> x != null && x != cls).toArray(Class[]::new);
        
        for (var field : cls.getFields())
            field.interfaceDeclarations = Arrays.stream(Arrays.stream(intfs).map(x -> Arrays.stream(x.getFields()).filter(f -> f.getName().equals(field.getName())).findFirst().orElse(null)).toArray(Field[]::new)).filter(x -> x != null).toArray(Field[]::new);
        
        for (var method : cls.getMethods()) {
            method.interfaceDeclarations = Arrays.stream(Arrays.stream(intfs).map(x -> Arrays.stream(x.getMethods()).filter(m -> m.name.equals(method.name)).findFirst().orElse(null)).toArray(Method[]::new)).filter(x -> x != null).toArray(Method[]::new);
            method.overrides = Arrays.stream(Arrays.stream(clses).map(x -> Arrays.stream(x.getMethods()).filter(m -> m.name.equals(method.name)).findFirst().orElse(null)).toArray(Method[]::new)).filter(x -> x != null).findFirst().orElse(null);
            if (method.overrides != null)
                method.overrides.overriddenBy.add(method);
        }
    }
    
    public void visitFiles(SourceFile[] files) {
        for (var file : files)
            for (var cls : file.classes)
                this.visitClass(cls);
    }
}